# <center>NLA Homework1</center>

### <center>Group O</center>

### <center>Dafu Zhu, Jiawei Ni</center>

## Problem 1

Find the value of x such that the matrix below has rank 3:
$$
\left(\begin{array}{llll}
x & 1 & 1 & 1 \\
1 & x & 1 & 1 \\
1 & 1 & x & 1 \\
1 & 1 & 1 & x
\end{array}\right) .
$$



*Solution:*

Given matrix:

$$
A = \begin{pmatrix} r_1 \\ r_2 \\ r_3 \\ r_4 \end{pmatrix} =
\begin{pmatrix} x & 1 & 1 & 1 \\ 1 & x & 1 & 1 \\ 1 & 1 & x & 1 \\ 1 & 1 & 1 & x \end{pmatrix}
$$

is row equivalent to:

$$
\begin{pmatrix} r_4 \\ r_2 \\ r_3 \\ r_1 \end{pmatrix} =
\begin{pmatrix} 1 & 1 & 1 & x \\ 1 & x & 1 & 1 \\ 1 & 1 & x & 1 \\ x & 1 & 1 & 1 \end{pmatrix}
$$

Performing row operations:

$$
\begin{pmatrix} r_4 \\ r_2 - r_4 \\ r_3 - r_4 \\ r_1 - x r_4 \end{pmatrix} =
\begin{pmatrix} 1 & 1 & 1 & x \\ 0 & x-1 & 0 & 1-x \\ 0 & 0 & x-1 & 1-x \\ 0 & 1-x & 1-x & 1-x^2 \end{pmatrix}
$$

$$
\begin{pmatrix} r_4 \\ r_2 - r_4 \\ r_3 - r_4 \\ r_1 + r_2 - (x+1) r_4 \end{pmatrix} =
\begin{pmatrix} 1 & 1 & 1 & x \\ 0 & x-1 & 0 & 1-x \\ 0 & 0 & x-1 & 1-x \\ 0 & 0 & 1-x & (1-x^2) + (1-x) \end{pmatrix}
$$

$$
\begin{pmatrix} r_4 \\ r_2 - r_4 \\ r_3 - r_4 \\ r_1 + r_2 + r_3 - (x+2) r_4 \end{pmatrix} =
\begin{pmatrix} 1 & 1 & 1 & x \\ 0 & x-1 & 0 & 1-x \\ 0 & 0 & x-1 & 1-x \\ 0 & 0 & 0 & (1-x^2) + 2(1-x) \end{pmatrix}
$$

where

$$
(1 - x^2) + 2(1-x) = -x^2 - 2x + 3 = -(x+3)(x-1)
$$

The determinant of A gives

$$
\det A = - (x+3)(x-1)^3 = 0
$$

Solving for $ x $:

$$
x_1 = 1, \quad x_2 = -3
$$

If $ x = 1 $, rank of $ A $ is 1.

If $ x = -3 $, rank of $ A $ is 3.

Thus, $ x = -3 $.

## Problem 2

Show that the product of two symmetric matrices is not necessarily a symmetric matrix.

Hint: Find $2 \times 2$ matrices $A$ and $B$ such that $A=A^{t}, B=B^{t}$, and $(A B)^{t} \neq A B$.

*Solution:*

Let:

$$
A = \begin{bmatrix} 1 & 3 \\ 3 & 2 \end{bmatrix}, \quad
B = \begin{bmatrix} 2 & 1 \\ 1 & 3 \end{bmatrix}
$$

$$
AB = \begin{bmatrix} 5 & 10 \\ 8 & 9 \end{bmatrix} \neq (AB)^T
$$

To generalize:

$$
A = \begin{bmatrix} a & c \\ c & b \end{bmatrix}, \quad
B = \begin{bmatrix} d & f \\ f & e \end{bmatrix}
$$

$$
AB = \begin{bmatrix} ad + cf & af + ce \\ cd + bf & cf + be \end{bmatrix}
$$

$ AB $ is symmetric if and only if:

$$
cd + bf = af + ce
$$

which simplifies to:

$$
c(d - e) = f(a - b)
$$

This condition does not always hold. Thus, the product of two symmetric matrices is not necessarily a symmetric matrix.

## Problem 3

Let $A$ be a square matrix of size $n$. Show that $A D=D A$ for any diagonal matrix $D$ of size $n$ if and only if $A$ is a diagonal matrix.

*Solution:*

(1) If $ A $ is diagonal, given $ D = \operatorname{diag}(d_k)_{k=1:n} $:

$$
A = \operatorname{diag}(a_k)_{k=1:n} = \operatorname{col}(a_k e_k)_{k=1:n}
$$

$$
AD = \operatorname{col}(d_k\cdot  a_k e_k)_{k=1:n} = \operatorname{diag}(d_k a_k)_{k=1:n}
$$

$$
DA = \operatorname{col}(a_k\cdot  d_k e_k)_{k=1:n} = \operatorname{diag}(a_k d_k)_{k=1:n}
$$

Thus, $ AD = DA $.

(2) If $ AD = DA $, 

Denote $A = \operatorname{col}(a_k)_{k=1:n} = \operatorname{row}(r_j)_{j=1:n}$

$D=\operatorname{col}(d_ke_k)_{k=1:n}=\operatorname{row}(d_je_j^t)_{j=1:n} $, where $e_i$ is the i-th column of the identity matrix

$$
AD(j, k) = r_j\cdot d_k e_k = d_k r_j(k)
$$

$$
DA(j, k) = d_j e_j \cdot a_k =d_j a_k(j)
$$

Since $ r_j(k) = a_k(j) = A(j, k) $, we have:

$$
\begin{split}
AD = DA & \Leftrightarrow AD(j, k) = DA(j, k)\\
 & \Leftrightarrow d_k \cdot A(j, k) = d_j \cdot A(j, k)\\
  & \Leftrightarrow (d_k-d_j)A(j,k)=0
\end{split}
$$

For all possible matrix $D$, $\forall j\neq k$, $d_k≠d_j$  $\therefore A(j,k)=0$, A is diagonal matrix

## Problem 4

Let $M_{1}, M_{2}, M_{3}$, and $M_{4}$ be square matrices of the same size such that $M_{1} M_{2} M_{3} M_{4}= I$. Show that

$$
M_{1} M_{2} M_{3} M_{4}=M_{2} M_{3} M_{4} M_{1}=M_{3} M_{4} M_{1} M_{2}=M_{4} M_{1} M_{2} M_{3}=I .
$$

*Solution:*

By taking inverses from both sides, we get

$$M_2M_3M_4=M_1^{-1}$$

$$M_3M_4=M_2^{-1}M_1^{-1}$$

$$M_4=M_3^{-1}M_2^{-1}M_1^{-1}$$

Since for any nonsingular matrix A, $A^{-1}A=AA^{-1}=I$

$$\therefore M_2M_3M_4M_1=M_1^{-1}M_1=I$$

$$M_3M_4M_1M_2=M_2^{-1}M_1^{-1}(M_1M_2)=M_2^{-1}M_2=I$$

$$M_4M_1M_2M_3=M_3^{-1}M_2^{-1}M_1^{-1}(M_1M_2M_3)=I$$

## Problem 5

Let $A$ be a nonsingular matrix. If $I+A$ and $I+A^{-1}$ are also nonsingular matrices, show that

$$
(I+A)^{-1}=I-\left(I+A^{-1}\right)^{-1} .
$$

*Solution:*

$I+A^{-1}=(I+A)A^{-1}$

$(I+A)^{-1}(I+A^{-1})=A^{-1}$ since $I+A$ is nonsingular

$(I+A)^{-1}(I+A^{-1})+I=A^{-1}+I$ (add $I$ on both sides)

$I+A^{-1}-(I+A)^{-1}(I+A^{-1})=I$

$(I-(I+A)^{-1})(I+A^{-1})=I$

$I-(I+A)^{-1}=(I+A^{-1})^{-1}$ since $(I+A^{-1})$ is nonsingular

$\therefore I-(I+A^{-1})^{-1}=(I+A)^{-1}$,  Q.E.D.

## Problem 6

Show that an $n \times n$ matrix $A$ has rank 1 if and only if there exist an $n \times 1$ nonzero column vector $v$ and an $1 \times n$ nonzero row vector $w^{t}$ such that $A=v w^{t}$

*Solution:*

Denote vectors $v=(u_i)_{i=1:n}$, $w=(w_i)_{i=1:n}$, not all zeros.

(1) If $A=vw^t=col(w_k \cdot v)_{k=1:n}$

All columns of A are multiples of vector $v$, which means all column vectors are linearly dependent, the column space of $A$ is spanned by the single vector $v$, $\Rightarrow rank(A)=1$

(2) If $rank(A)=1$, then the column space of $A$ is spanned by a single nonzero vector $v$. $\Rightarrow$ every column of $A$ is a multiple of vector $v$

$\exists w_i\in \mathbb{R}, i=1:n, s.t.$ $A=(w_1v | \cdots | w_nv)=v(w_1 | \cdots | w_n)=vw^t$

Since $A$ is nonzero (or else the rank would be 0), then $v$ and $w$ are both nonzero column vectors.

## Problem 7

Let $x$ and $y$ be column vectors of size $n$, and let $I$ be the identity matrix of size $n$. Show that the matrix $I-x y^{t}$ is nonsingular if and only if $y^{t} x \neq 1$.

*Solution:*

(1) If $y^tx\neq 1$, we show that $I-xy^t$ is nonsingular

cf. derivation of Sherman-Morrison-Woodbury formula:

Solve $(I-xy^t)v=b$

$v-xy^tv=b$, denote $\xi = y^tv$

Consider the bordered matrix

$$\begin{bmatrix} I & -x \\ y^t & -1 \end{bmatrix} \begin{bmatrix} v \\ \xi \end{bmatrix} = \begin{bmatrix} b \\ 0 \end{bmatrix}$$

Factorizing

$$
\begin{bmatrix} I & -x \\ y^t & -1 \end{bmatrix} = \begin{bmatrix} I & 0 \\ y^t & L_{22} \end{bmatrix} \begin{bmatrix} I & -x \\ 0 & U_{22} \end{bmatrix}
$$

$L_{22}U_{22} = y^tx - 1$

Consider intermediate variable $\begin{bmatrix} y_1 \\ y_2 \end{bmatrix}$, s.t. $\begin{bmatrix} I & 0 \\ y^t & L_{22} \end{bmatrix} \begin{bmatrix} y_1 \\ y_2 \end{bmatrix} = \begin{bmatrix} b \\ 0 \end{bmatrix}$

$$
y_1 = b \tag{1}
$$

$$
y^ty_1 + L_{22}y_2 = 0 \Rightarrow y_2 = -L_{22}^{-1}y^ty_1  \tag{2}
$$

$$
\begin{bmatrix} I & -x \\ 0 & U_{22} \end{bmatrix} \begin{bmatrix} v \\ \xi \end{bmatrix} = \begin{bmatrix} y_1 \\ y_2 \end{bmatrix}
$$

$$
v=y_1+x\xi \tag{3}
$$
$$
\xi = U_{22}^{-1}y_2 \tag{4}
$$

By equations $(1)(2)(3)(4)$, 

$$
\begin{split}
v &= b + xU_{22}^{-1}y_2\\
&=b - xU_{22}^{-1}L_{22}^{-1}y^ty_1\\
&=b - \frac{1}{y^tx - 1}xy^tb\\
&=\left(I + \frac{1}{1 - y^tx}xy^t\right)b
\end{split}
$$

Since $y^tx \neq 1$, 

$$(I - xy^t)^{-1} = I + \frac{1}{1 - y^tx}xy^t$$

(2) Now we only need to show that if $y^tx = 1$ then the matrix $I - xy^t$ is singular.

Assume that $y^tx = 1$

$(I - xy^t)x = x - xy^tx = x - x = 0$

If $I - xy^t$ is nonsingular, multiply its inverse from both sides,

$(I - xy^t)^{-1}(I - xy^t)x = 0 \Rightarrow x = 0$

This is contradictory to $y^tx = 1$

$\therefore I - xy^t$ is singular

Therefore $I-x y^{t}$ is nonsingular if and only if $y^{t} x \neq 1$, Q.E.D.

## Problem 8

The covariance matrix of five random variables is

$$
\Sigma=\left(\begin{array}{ccccc}
1 & -0.525 & 1.375 & -0.075 & -0.75 \\
-0.525 & 4 & 0.1875 & 0.1875 & -0.675 \\
1.375 & 0.1875 & 12.25 & 0.4375 & -1.875 \\
-0.075 & 0.1875 & 0.4375 & 6.25 & 0.3 \\
-0.75 & -0.675 & -1.875 & 0.3 & 4.41
\end{array}\right)
$$

Find the correlation matrix of these random variables.

*Solution:*

$\Omega_x = D_{\sigma}^{-1} \Sigma D_{\sigma}^{-1}$

$D_{\sigma} = \text{diag}(\sqrt{\sum(i,i)})_{i=1:5}$

$D_{\sigma}^{-1} = \text{diag}\left(\frac{1}{\sqrt{\sum(i,i)}}\right)_{i=1:5}$

$D_{\sigma}^{-1} = \begin{pmatrix} 1 & 0 & 0 & -0 & 0 \\ 0 & 1/2 & 0 & 0 & 0 \\ 0 & 0 & 1/3.5 & 0 & 0 \\ 0 & 0 & 0 & 1/2.5 & 0 \\ 0 & 0 & 0 & 0 & 1/2.1 \end{pmatrix}$

$\Omega_x = D_{\sigma}^{-1} \Sigma D_{\sigma}^{-1}= \begin{pmatrix}
1 & -0.2625 & 0.392857 & -0.03 & -0.357143 \\
-0.2625 & 1 & 0.026786 & 0.0375 & -0.160714 \\
0.392857 & 0.026786 & 1 & 0.05 & -0.255102 \\
-0.03 & 0.0375 & 0.05 & 1 & 0.057143 \\
-0.357143 & -0.160714 & -0.255102 & 0.057143 & 1
\end{pmatrix}$

In [7]:
import numpy as np
import pandas as pd
np.set_printoptions(suppress=True, precision=6)  # Round my answers to 6 decimal places

In [4]:
# (8)
sig_inv = np.diag([1, 1/2, 1/3.5, 1/2.5, 1/2.1])
l = np.array([
    [1, 0, 0, 0, 0],
    [-0.525, 4, 0, 0, 0],
    [1.375, 0.1875, 12.25, 0, 0],
    [-0.075, 0.1875, 0.4375, 6.25, 0],
    [-0.75, -0.675, -1.875, 0.3, 4.41]
])
Sig = l + np.tril(l, -1).T
Ome = sig_inv @ Sig @ sig_inv
Ome

array([[ 1.      , -0.2625  ,  0.392857, -0.03    , -0.357143],
       [-0.2625  ,  1.      ,  0.026786,  0.0375  , -0.160714],
       [ 0.392857,  0.026786,  1.      ,  0.05    , -0.255102],
       [-0.03    ,  0.0375  ,  0.05    ,  1.      ,  0.057143],
       [-0.357143, -0.160714, -0.255102,  0.057143,  1.      ]])

## Problem 9

The correlation matrix of five random variables is

$$
\Omega=\left(\begin{array}{ccccc}
1 & -0.25 & 0.15 & -0.05 & -0.30 \\
-0.25 & 1 & -0.10 & -0.25 & 0.10 \\
0.15 & -0.10 & 1 & 0.20 & 0.05 \\
-0.05 & -0.25 & 0.20 & 1 & 0.10 \\
-0.30 & 0.10 & 0.05 & 0.10 & 1
\end{array}\right)
$$

(i) Compute the covariance matrix of these random variables if their standard deviations are $0.1,0.2,0.5,1$, and 2 , in this order.

(ii) Compute the covariance matrix of these random variables if their standard deviations are $2,1,0.5,0.2$, and 0.1 , in this order.

*Solution:*

(i) $\Sigma=D_{\sigma}\Omega D_{\sigma}$, $D_{\sigma}=\text{diag}(0.1,0.2,0.5,1,2)$

$$
\Sigma=D_{\sigma}\Omega D_{\sigma}=\begin{pmatrix}
0.01 & -0.005 & 0.0075 & -0.005 & -0.06 \\
-0.005 & 0.04 & -0.01 & -0.05 & 0.04 \\
0.0075 & -0.01 & 0.25 & 0.1 & 0.05 \\
-0.005 & -0.05 & 0.1 & 1 & 0.2 \\
-0.06 & 0.04 & 0.05 & 0.2 & 4
\end{pmatrix}
$$

(ii) $D_{\sigma_2}=\text{diag}(2, 1, 0.5, 0.2, 0.1)$

$$
\Sigma_2=D_{\sigma_2}\Omega D_{\sigma_2}=\begin{pmatrix}
4 & -0.5 & 0.15 & -0.02 & -0.06 \\
-0.5 & 1 & -0.05 & -0.05 & 0.01 \\
0.15 & -0.05 & 0.25 & 0.02 & 0.0025 \\
-0.02 & -0.05 & 0.02 & 0.04 & 0.002 \\
-0.06 & 0.01 & 0.0025 & 0.002 & 0.01
\end{pmatrix}
$$

In [6]:
# (9)
## (i)
D_sig = np.diag([0.1, 0.2, 0.5, 1, 2])
l = np.array([
    [1, 0, 0, 0, 0],
    [-0.25, 1, 0, 0, 0],
    [0.15, -0.1, 1, 0, 0],
    [-0.05, -0.25, 0.2, 1, 0],
    [-0.3, 0.1, 0.05, 0.1, 1]
])
Ome = l + np.tril(l, -1).T
a1 = D_sig @ Ome @ D_sig
print("(i)")
print(a1)

## (ii)
D_sig2 = np.diag([2, 1, 0.5, 0.2, 0.1])
a2 = D_sig2 @ Ome @ D_sig2
print("(ii)")
print(a2)

(i)
[[ 0.01   -0.005   0.0075 -0.005  -0.06  ]
 [-0.005   0.04   -0.01   -0.05    0.04  ]
 [ 0.0075 -0.01    0.25    0.1     0.05  ]
 [-0.005  -0.05    0.1     1.      0.2   ]
 [-0.06    0.04    0.05    0.2     4.    ]]
(ii)
[[ 4.     -0.5     0.15   -0.02   -0.06  ]
 [-0.5     1.     -0.05   -0.05    0.01  ]
 [ 0.15   -0.05    0.25    0.02    0.0025]
 [-0.02   -0.05    0.02    0.04    0.002 ]
 [-0.06    0.01    0.0025  0.002   0.01  ]]


## Problem 10

The file indeces-close-jan3-jan31-2017.xlsx contains the January 3, 2017 - January 31, 2017 end of day values of Dow Jones, Nasdaq, and S\&P 500.

(i) Compute the log daily returns of the three indices over the given time period.

(ii) Compute the sample covariance matrix of the log daily returns of the three indices over the given time period.

(iii) Compute the percentage daily returns of the three indices over the given time period.

(iv) Compute the sample covariance matrix of the percentage daily returns of the three indices over the given time period.

In [8]:
# import data
close = pd.read_excel("data/indeces-close-jan3-jan31-2017.xlsx", index_col=0)
close.head()

DOW       NASDAQ        SP500
Date                                             
2017-01-03  19881.75977  5429.080078  2257.830078
2017-01-04  19942.16016  5477.000000  2270.750000
2017-01-05  19899.28906  5487.939941  2269.000000
2017-01-06  19963.80078  5521.060059  2276.979980
2017-01-09  19887.38086  5531.819824  2268.899902

*Solutions:*

(i) Compute the log daily returns of the three indices over the given time period.

$$
r = P_2/P_1 - 1
$$

$$
\text{log\_return} = \log(P_2/P_1) = \log(r+1)
$$

Denote the matrix `log_return` as $T_X$

$$
T_X =\begin{pmatrix}
0.003033 & 0.008788 & 0.005706 \\
-0.002152 & 0.001995 & -0.000771 \\
0.003237 & 0.006017 & 0.003511 \\
-0.003835 & 0.001947 & -0.003555 \\
-0.001603 & 0.003609 & 0.0 \\
0.004961 & 0.002129 & 0.002826 \\
-0.003176 & -0.002909 & -0.002147 \\
-0.000265 & 0.004789 & 0.001848 \\
-0.002969 & -0.006369 & -0.002972 \\
-0.001113 & 0.00305 & 0.001762 \\
-0.003658 & -0.002806 & -0.003616 \\
0.004795 & 0.002749 & 0.003361 \\
-0.001383 & -0.00043 & -0.002694 \\
0.005684 & 0.00861 & 0.006543 \\
0.007794 & 0.009839 & 0.007994 \\
0.001613 & -0.000205 & -0.000736 \\
-0.000355 & 0.00099 & -0.000867 \\
-0.006123 & -0.00835 & -0.006028 \\
-0.005374 & 0.000192 & -0.00089
\end{pmatrix}
$$

In [9]:
def get_lr(df):
    """
    get log return
    """
    log_return = np.log((df.pct_change().dropna() + 1).values)
    return log_return

log_return = get_lr(close)
log_return

array([[ 0.003033,  0.008788,  0.005706],
       [-0.002152,  0.001995, -0.000771],
       [ 0.003237,  0.006017,  0.003511],
       [-0.003835,  0.001947, -0.003555],
       [-0.001603,  0.003609,  0.      ],
       [ 0.004961,  0.002129,  0.002826],
       [-0.003176, -0.002909, -0.002147],
       [-0.000265,  0.004789,  0.001848],
       [-0.002969, -0.006369, -0.002972],
       [-0.001113,  0.00305 ,  0.001762],
       [-0.003658, -0.002806, -0.003616],
       [ 0.004795,  0.002749,  0.003361],
       [-0.001383, -0.00043 , -0.002694],
       [ 0.005684,  0.00861 ,  0.006543],
       [ 0.007794,  0.009839,  0.007994],
       [ 0.001613, -0.000205, -0.000736],
       [-0.000355,  0.00099 , -0.000867],
       [-0.006123, -0.00835 , -0.006028],
       [-0.005374,  0.000192, -0.00089 ]])

(ii) Compute the sample covariance matrix of the log daily returns of the three indices
over the given time period.

Recall that

$$
\bar{T}_{X_k} = T_{X_k} - \mu_{X_k}
$$

$$
\bar{T}_X=\text{col} (\bar{T}_{X_k})_{k=1:n}
$$

By (1.37) in the textbook, the sample covariance matrix

$$
\hat{\Sigma}_X=\frac{1}{N-1} \bar{T}_X^t\bar{T}_X=\begin{pmatrix}
0.000016 & 0.000015 & 0.000014 \\
0.000015 & 0.000023 & 0.000016 \\
0.000014 & 0.000016 & 0.000014
\end{pmatrix}
$$

In [19]:
def get_Sig(T):
    """
    get the sample cov Sigma given matrix T
    """
    T_bar = T - T.mean(axis=0)
    Sig = 1 / (T_bar.shape[0] - 1) * T_bar.T @ T_bar
    return Sig

T = log_return
Sig = get_Sig(T)
Sig

array([[0.000016, 0.000015, 0.000014],
       [0.000015, 0.000023, 0.000016],
       [0.000014, 0.000016, 0.000014]])

(iii) Compute the percentage daily returns of the three indices over the given time
period.

$$
r=P_2/P_1-1
$$

$$
T_X=\begin{pmatrix}
0.003038 & 0.008827 & 0.005722 \\
-0.00215 & 0.001997 & -0.000771 \\
0.003242 & 0.006035 & 0.003517 \\
-0.003828 & 0.001949 & -0.003549 \\
-0.001602 & 0.003615 & 0.0 \\
0.004973 & 0.002131 & 0.00283 \\
-0.003171 & -0.002905 & -0.002145 \\
-0.000265 & 0.0048 & 0.00185 \\
-0.002965 & -0.006349 & -0.002968 \\
-0.001112 & 0.003055 & 0.001764 \\
-0.003652 & -0.002803 & -0.003609 \\
0.004807 & 0.002753 & 0.003366 \\
-0.001382 & -0.00043 & -0.00269 \\
0.0057 & 0.008648 & 0.006565 \\
0.007824 & 0.009888 & 0.008026 \\
0.001614 & -0.000205 & -0.000735 \\
-0.000355 & 0.00099 & -0.000866 \\
-0.006104 & -0.008315 & -0.00601 \\
-0.00536 & 0.000192 & -0.00089
\end{pmatrix}
$$

In [20]:
def get_pctchg(df):
    """
    get percentage return
    """
    pct_chg = df.pct_change().dropna().values
    return pct_chg

pct_chg = get_pctchg(close)
pct_chg

array([[ 0.003038,  0.008827,  0.005722],
       [-0.00215 ,  0.001997, -0.000771],
       [ 0.003242,  0.006035,  0.003517],
       [-0.003828,  0.001949, -0.003549],
       [-0.001602,  0.003615,  0.      ],
       [ 0.004973,  0.002131,  0.00283 ],
       [-0.003171, -0.002905, -0.002145],
       [-0.000265,  0.0048  ,  0.00185 ],
       [-0.002965, -0.006349, -0.002968],
       [-0.001112,  0.003055,  0.001764],
       [-0.003652, -0.002803, -0.003609],
       [ 0.004807,  0.002753,  0.003366],
       [-0.001382, -0.00043 , -0.00269 ],
       [ 0.0057  ,  0.008648,  0.006565],
       [ 0.007824,  0.009888,  0.008026],
       [ 0.001614, -0.000205, -0.000735],
       [-0.000355,  0.00099 , -0.000866],
       [-0.006104, -0.008315, -0.00601 ],
       [-0.00536 ,  0.000192, -0.00089 ]])

(iv) Compute the sample covariance matrix of the percentage daily returns of the three
indices over the given time period.

$$
\hat{\Sigma}_X=\frac{1}{N-1} \bar{T}_X^t\bar{T}_X=\begin{pmatrix}
0.000016 & 0.000015 & 0.000014 \\
0.000015 & 0.000023 & 0.000017 \\
0.000014 & 0.000017 & 0.000014
\end{pmatrix}
$$

In [21]:
T2 = pct_chg
Sig2 = get_Sig(T2)
Sig2

array([[0.000016, 0.000015, 0.000014],
       [0.000015, 0.000023, 0.000017],
       [0.000014, 0.000017, 0.000014]])

## Problem 11

The file indices-july2016.csv contains the January 2016 - July 2016 end of day values of nine major US indeces.

(i) Compute the sample covariance matrix of the daily percentage returns of the indeces, and the corresponding sample corelation matrix.

Compute the sample covariance and correlation matrices for daily log returns, and compare them with the corresponding matrices for daily percentage returns.

(ii) Compute the sample covariance matrix of the weekly percentage returns of the indeces, and the corresponding sample corelation matrix.

Compute the sample covariance and correlation matrices for weekly log returns, and compare them with the corresponding matrices for weekly percentage returns.

(iii) Compute the sample covariance matrix of the monthly percentage returns of the indeces, and the corresponding sample corelation matrix.

Compute the sample covariance and correlation matrices for monthly log returns, and compare them with the corresponding matrices for monthly percentage returns.

(iv) Comment on the differences between the sample covariance and correlation matrices for daily, weekly, and monthly returns.

*Solutions:*

In [22]:
# import data
value = pd.read_csv("data/indices-july2016.csv", index_col=0)
value.index = pd.to_datetime(value.index)
value.head()

BAC        BNS          C          GS        JPM         MS  \
Date                                                                            
2016-01-04  16.176800  38.183614  50.516574  174.428759  61.870034  30.587742   
2016-01-05  16.176800  38.451837  50.249813  171.425438  61.977008  30.393412   
2016-01-06  15.832194  37.522631  49.518689  167.240487  61.082316  29.635520   
2016-01-07  15.261132  36.909548  46.989405  162.100381  58.612182  28.158602   
2016-01-08  14.965755  36.928707  45.576561  161.430796  57.299313  27.575607   

                   TD        USB        WFC  
Date                                         
2016-01-04  36.790199  40.463358  50.916643  
2016-01-05  36.503965  40.658457  50.897396  
2016-01-06  35.689043  40.043897  49.925449  
2016-01-07  34.889533  39.283015  48.501207  
2016-01-08  34.735410  38.726986  47.692854

Compact the covariance matrix and correlation matrix of log returns and percentage returns into functions.

In [23]:
def cov_log(df):
    """
    covariance matrix of log returns
    """
    T = get_lr(df)
    cov = get_Sig(T)
    return cov

def corr_log(df):
    """
    correlation matrix of log returns
    """
    Sig = cov_log(df)
    D_sig = np.diag(np.sqrt(np.diag(Sig)))
    D_sig_inv = np.linalg.inv(D_sig)
    corr = D_sig_inv @ Sig @ D_sig_inv
    return corr

def cov_pct(df):
    """
    covariance matrix of percentage returns
    """
    T = get_pctchg(df)
    cov = get_Sig(T)
    return cov

def corr_pct(df):
    """
    correlation matrix of percentage returns
    """
    Sig = cov_pct(df)
    D_sig = np.diag(np.sqrt(np.diag(Sig)))
    D_sig_inv = np.linalg.inv(D_sig)
    corr = D_sig_inv @ Sig @ D_sig_inv
    return corr

(i) Compute the sample covariance matrix of the daily percentage returns of the indeces, and the corresponding sample corelation matrix.

Compute the sample covariance and correlation matrices for daily log returns, and
compare them with the corresponding matrices for daily percentage returns.

Sample covariance matrix of daily percentage returns:

$$
\begin{pmatrix}
0.000531 & 0.000246 & 0.000521 & 0.000375 & 0.000377 & 0.000496 & 0.000196 & 0.000293 & 0.000295 \\
0.000246 & 0.000303 & 0.000274 & 0.000212 & 0.00021 & 0.000265 & 0.000217 & 0.000169 & 0.000164 \\
0.000521 & 0.000274 & 0.000574 & 0.000399 & 0.000402 & 0.000532 & 0.000224 & 0.000309 & 0.000313 \\
0.000375 & 0.000212 & 0.000399 & 0.00035 & 0.000294 & 0.000411 & 0.000174 & 0.000238 & 0.000232 \\
0.000377 & 0.00021 & 0.000402 & 0.000294 & 0.000326 & 0.000387 & 0.000166 & 0.000245 & 0.000235 \\
0.000496 & 0.000265 & 0.000532 & 0.000411 & 0.000387 & 0.000582 & 0.000222 & 0.00031 & 0.000305 \\
0.000196 & 0.000217 & 0.000224 & 0.000174 & 0.000166 & 0.000222 & 0.000193 & 0.000139 & 0.000135 \\
0.000293 & 0.000169 & 0.000309 & 0.000238 & 0.000245 & 0.00031 & 0.000139 & 0.000222 & 0.000197 \\
0.000295 & 0.000164 & 0.000313 & 0.000232 & 0.000235 & 0.000305 & 0.000135 & 0.000197 & 0.000226
\end{pmatrix}
$$

In [24]:
daily = value
print("Sample covariance matrix of daily percentage returns:")
pct_day_cov = pd.DataFrame(cov_pct(daily))
pct_day_cov

Sample covariance matrix of daily percentage returns:


0         1         2         3         4         5         6  \
0  0.000531  0.000246  0.000521  0.000375  0.000377  0.000496  0.000196   
1  0.000246  0.000303  0.000274  0.000212  0.000210  0.000265  0.000217   
2  0.000521  0.000274  0.000574  0.000399  0.000402  0.000532  0.000224   
3  0.000375  0.000212  0.000399  0.000350  0.000294  0.000411  0.000174   
4  0.000377  0.000210  0.000402  0.000294  0.000326  0.000387  0.000166   
5  0.000496  0.000265  0.000532  0.000411  0.000387  0.000582  0.000222   
6  0.000196  0.000217  0.000224  0.000174  0.000166  0.000222  0.000193   
7  0.000293  0.000169  0.000309  0.000238  0.000245  0.000310  0.000139   
8  0.000295  0.000164  0.000313  0.000232  0.000235  0.000305  0.000135   

          7         8  
0  0.000293  0.000295  
1  0.000169  0.000164  
2  0.000309  0.000313  
3  0.000238  0.000232  
4  0.000245  0.000235  
5  0.000310  0.000305  
6  0.000139  0.000135  
7  0.000222  0.000197  
8  0.000197  0.000226

Sample correlation matrix of daily percentage returns:

$$
\begin{pmatrix}
1.0 & 0.614084 & 0.943582 & 0.869843 & 0.905068 & 0.893289 & 0.612314 & 0.853221 & 0.852008 \\
0.614084 & 1.0 & 0.656837 & 0.649994 & 0.669019 & 0.631424 & 0.8956 & 0.651481 & 0.625134 \\
0.943582 & 0.656837 & 1.0 & 0.891605 & 0.928207 & 0.9215 & 0.672348 & 0.865484 & 0.868332 \\
0.869843 & 0.649994 & 0.891605 & 1.0 & 0.869316 & 0.911971 & 0.669761 & 0.853871 & 0.824126 \\
0.905068 & 0.669019 & 0.928207 & 0.869316 & 1.0 & 0.88855 & 0.66242 & 0.90917 & 0.86572 \\
0.893289 & 0.631424 & 0.9215 & 0.911971 & 0.88855 & 1.0 & 0.663019 & 0.860713 & 0.842499 \\
0.612314 & 0.8956 & 0.672348 & 0.669761 & 0.66242 & 0.663019 & 1.0 & 0.670164 & 0.647531 \\
0.853221 & 0.651481 & 0.865484 & 0.853871 & 0.90917 & 0.860713 & 0.670164 & 1.0 & 0.879214 \\
0.852008 & 0.625134 & 0.868332 & 0.824126 & 0.86572 & 0.842499 & 0.647531 & 0.879214 & 1.0
\end{pmatrix}
$$

In [27]:
print("Sample correlation matrix of daily percentage returns:")
pct_day_corr = pd.DataFrame(corr_pct(daily))
pct_day_corr

Sample correlation matrix of daily percentage returns:


0         1         2         3         4         5         6  \
0  1.000000  0.614084  0.943582  0.869843  0.905068  0.893289  0.612314   
1  0.614084  1.000000  0.656837  0.649994  0.669019  0.631424  0.895600   
2  0.943582  0.656837  1.000000  0.891605  0.928207  0.921500  0.672348   
3  0.869843  0.649994  0.891605  1.000000  0.869316  0.911971  0.669761   
4  0.905068  0.669019  0.928207  0.869316  1.000000  0.888550  0.662420   
5  0.893289  0.631424  0.921500  0.911971  0.888550  1.000000  0.663019   
6  0.612314  0.895600  0.672348  0.669761  0.662420  0.663019  1.000000   
7  0.853221  0.651481  0.865484  0.853871  0.909170  0.860713  0.670164   
8  0.852008  0.625134  0.868332  0.824126  0.865720  0.842499  0.647531   

          7         8  
0  0.853221  0.852008  
1  0.651481  0.625134  
2  0.865484  0.868332  
3  0.853871  0.824126  
4  0.909170  0.865720  
5  0.860713  0.842499  
6  0.670164  0.647531  
7  1.000000  0.879214  
8  0.879214  1.000000

Sample covariance matrix of daily log returns:

$$
\begin{pmatrix}
0.000535 & 0.000247 & 0.000526 & 0.000379 & 0.000378 & 0.000502 & 0.000198 & 0.000296 & 0.000296 \\
0.000247 & 0.0003 & 0.000274 & 0.000212 & 0.000209 & 0.000266 & 0.000216 & 0.000169 & 0.000163 \\
0.000526 & 0.000274 & 0.00058 & 0.000405 & 0.000403 & 0.00054 & 0.000226 & 0.000312 & 0.000314 \\
0.000379 & 0.000212 & 0.000405 & 0.000354 & 0.000295 & 0.000417 & 0.000176 & 0.000241 & 0.000233 \\
0.000378 & 0.000209 & 0.000403 & 0.000295 & 0.000325 & 0.00039 & 0.000166 & 0.000246 & 0.000235 \\
0.000502 & 0.000266 & 0.00054 & 0.000417 & 0.00039 & 0.00059 & 0.000224 & 0.000313 & 0.000307 \\
0.000198 & 0.000216 & 0.000226 & 0.000176 & 0.000166 & 0.000224 & 0.000193 & 0.00014 & 0.000136 \\
0.000296 & 0.000169 & 0.000312 & 0.000241 & 0.000246 & 0.000313 & 0.00014 & 0.000224 & 0.000198 \\
0.000296 & 0.000163 & 0.000314 & 0.000233 & 0.000235 & 0.000307 & 0.000136 & 0.000198 & 0.000226
\end{pmatrix}
$$

In [28]:
print("Sample covariance matrix of daily log returns:")
log_day_cov = pd.DataFrame(cov_log(daily))
log_day_cov

Sample covariance matrix of daily log returns:


0         1         2         3         4         5         6  \
0  0.000535  0.000247  0.000526  0.000379  0.000378  0.000502  0.000198   
1  0.000247  0.000300  0.000274  0.000212  0.000209  0.000266  0.000216   
2  0.000526  0.000274  0.000580  0.000405  0.000403  0.000540  0.000226   
3  0.000379  0.000212  0.000405  0.000354  0.000295  0.000417  0.000176   
4  0.000378  0.000209  0.000403  0.000295  0.000325  0.000390  0.000166   
5  0.000502  0.000266  0.000540  0.000417  0.000390  0.000590  0.000224   
6  0.000198  0.000216  0.000226  0.000176  0.000166  0.000224  0.000193   
7  0.000296  0.000169  0.000312  0.000241  0.000246  0.000313  0.000140   
8  0.000296  0.000163  0.000314  0.000233  0.000235  0.000307  0.000136   

          7         8  
0  0.000296  0.000296  
1  0.000169  0.000163  
2  0.000312  0.000314  
3  0.000241  0.000233  
4  0.000246  0.000235  
5  0.000313  0.000307  
6  0.000140  0.000136  
7  0.000224  0.000198  
8  0.000198  0.000226

Sample correlation matrix of daily log returns:

$$
\begin{pmatrix}
1.0 & 0.615775 & 0.943308 & 0.870803 & 0.905344 & 0.893048 & 0.614656 & 0.853932 & 0.850373 \\
0.615775 & 1.0 & 0.658394 & 0.65206 & 0.670548 & 0.632548 & 0.896778 & 0.652902 & 0.626568 \\
0.943308 & 0.658394 & 1.0 & 0.893366 & 0.928466 & 0.922524 & 0.676455 & 0.866574 & 0.867723 \\
0.870803 & 0.65206 & 0.893366 & 1.0 & 0.871169 & 0.912862 & 0.671991 & 0.854523 & 0.824505 \\
0.905344 & 0.670548 & 0.928466 & 0.871169 & 1.0 & 0.889834 & 0.664224 & 0.911052 & 0.865464 \\
0.893048 & 0.632548 & 0.922524 & 0.912862 & 0.889834 & 1.0 & 0.665056 & 0.860777 & 0.842024 \\
0.614656 & 0.896778 & 0.676455 & 0.671991 & 0.664224 & 0.665056 & 1.0 & 0.671095 & 0.649924 \\
0.853932 & 0.652902 & 0.866574 & 0.854523 & 0.911052 & 0.860777 & 0.671095 & 1.0 & 0.878271 \\
0.850373 & 0.626568 & 0.867723 & 0.824505 & 0.865464 & 0.842024 & 0.649924 & 0.878271 & 1.0
\end{pmatrix}
$$

In [32]:
print("Sample correlation matrix of daily log returns:")
log_day_corr = pd.DataFrame(corr_log(daily))
log_day_corr

Sample correlation matrix of daily log returns:


0         1         2         3         4         5         6  \
0  1.000000  0.615775  0.943308  0.870803  0.905344  0.893048  0.614656   
1  0.615775  1.000000  0.658394  0.652060  0.670548  0.632548  0.896778   
2  0.943308  0.658394  1.000000  0.893366  0.928466  0.922524  0.676455   
3  0.870803  0.652060  0.893366  1.000000  0.871169  0.912862  0.671991   
4  0.905344  0.670548  0.928466  0.871169  1.000000  0.889834  0.664224   
5  0.893048  0.632548  0.922524  0.912862  0.889834  1.000000  0.665056   
6  0.614656  0.896778  0.676455  0.671991  0.664224  0.665056  1.000000   
7  0.853932  0.652902  0.866574  0.854523  0.911052  0.860777  0.671095   
8  0.850373  0.626568  0.867723  0.824505  0.865464  0.842024  0.649924   

          7         8  
0  0.853932  0.850373  
1  0.652902  0.626568  
2  0.866574  0.867723  
3  0.854523  0.824505  
4  0.911052  0.865464  
5  0.860777  0.842024  
6  0.671095  0.649924  
7  1.000000  0.878271  
8  0.878271  1.000000

**Comparisons**



- Overall, the differences between covariance matrices and correlation matrices of daily percentage returns and log returns are very close to each other

(ii) Compute the sample covariance matrix of the weekly percentage returns of the
indeces, and the corresponding sample corelation matrix.

Compute the sample covariance and correlation matrices for weekly log returns, and compare them with the corresponding matrices for weekly percentage returns.

Sample covariance matrix of weekly percentage returns:

$$
\begin{pmatrix}
0.002115 & 0.000916 & 0.001868 & 0.001289 & 0.000995 & 0.001771 & 0.000572 & 0.000795 & 0.000852 \\
0.000916 & 0.001775 & 0.001312 & 0.001053 & 0.000726 & 0.001018 & 0.001058 & 0.00067 & 0.000575 \\
0.001868 & 0.001312 & 0.002134 & 0.001504 & 0.001197 & 0.001886 & 0.00084 & 0.000904 & 0.000893 \\
0.001289 & 0.001053 & 0.001504 & 0.001407 & 0.000867 & 0.001492 & 0.00075 & 0.000771 & 0.000682 \\
0.000995 & 0.000726 & 0.001197 & 0.000867 & 0.000789 & 0.001033 & 0.000472 & 0.000568 & 0.000549 \\
0.001771 & 0.001018 & 0.001886 & 0.001492 & 0.001033 & 0.001992 & 0.000765 & 0.000833 & 0.000843 \\
0.000572 & 0.001058 & 0.00084 & 0.00075 & 0.000472 & 0.000765 & 0.000774 & 0.00043 & 0.000371 \\
0.000795 & 0.00067 & 0.000904 & 0.000771 & 0.000568 & 0.000833 & 0.00043 & 0.000542 & 0.000442 \\
0.000852 & 0.000575 & 0.000893 & 0.000682 & 0.000549 & 0.000843 & 0.000371 & 0.000442 & 0.000598
\end{pmatrix}
$$

In [33]:
weekly = value.resample("W").last()
print("Sample covariance matrix of weekly percentage returns:")
pct_week_cov = pd.DataFrame(cov_pct(weekly))
pct_week_cov

Sample covariance matrix of weekly percentage returns:


0         1         2         3         4         5         6  \
0  0.002115  0.000916  0.001868  0.001289  0.000995  0.001771  0.000572   
1  0.000916  0.001775  0.001312  0.001053  0.000726  0.001018  0.001058   
2  0.001868  0.001312  0.002134  0.001504  0.001197  0.001886  0.000840   
3  0.001289  0.001053  0.001504  0.001407  0.000867  0.001492  0.000750   
4  0.000995  0.000726  0.001197  0.000867  0.000789  0.001033  0.000472   
5  0.001771  0.001018  0.001886  0.001492  0.001033  0.001992  0.000765   
6  0.000572  0.001058  0.000840  0.000750  0.000472  0.000765  0.000774   
7  0.000795  0.000670  0.000904  0.000771  0.000568  0.000833  0.000430   
8  0.000852  0.000575  0.000893  0.000682  0.000549  0.000843  0.000371   

          7         8  
0  0.000795  0.000852  
1  0.000670  0.000575  
2  0.000904  0.000893  
3  0.000771  0.000682  
4  0.000568  0.000549  
5  0.000833  0.000843  
6  0.000430  0.000371  
7  0.000542  0.000442  
8  0.000442  0.000598

Sample correlation matrix of weekly percentage returns:

$$
\begin{pmatrix}
1.0 & 0.472693 & 0.879529 & 0.747518 & 0.77015 & 0.862662 & 0.446917 & 0.742043 & 0.757836 \\
0.472693 & 1.0 & 0.674238 & 0.666234 & 0.613691 & 0.541635 & 0.902576 & 0.683105 & 0.558006 \\
0.879529 & 0.674238 & 1.0 & 0.868409 & 0.92265 & 0.914616 & 0.65394 & 0.840487 & 0.790964 \\
0.747518 & 0.666234 & 0.868409 & 1.0 & 0.82321 & 0.891044 & 0.718928 & 0.883104 & 0.744359 \\
0.77015 & 0.613691 & 0.92265 & 0.82321 & 1.0 & 0.823958 & 0.60445 & 0.867464 & 0.799954 \\
0.862662 & 0.541635 & 0.914616 & 0.891044 & 0.823958 & 1.0 & 0.616434 & 0.801593 & 0.772431 \\
0.446917 & 0.902576 & 0.65394 & 0.718928 & 0.60445 & 0.616434 & 1.0 & 0.663532 & 0.545015 \\
0.742043 & 0.683105 & 0.840487 & 0.883104 & 0.867464 & 0.801593 & 0.663532 & 1.0 & 0.775686 \\
0.757836 & 0.558006 & 0.790964 & 0.744359 & 0.799954 & 0.772431 & 0.545015 & 0.775686 & 1.0
\end{pmatrix}
$$

In [38]:
print("Sample correlation matrix of weekly percentage returns:")
pct_week_corr = pd.DataFrame(corr_pct(weekly))
pct_week_corr

Sample correlation matrix of weekly percentage returns:


0         1         2         3         4         5         6  \
0  1.000000  0.472693  0.879529  0.747518  0.770150  0.862662  0.446917   
1  0.472693  1.000000  0.674238  0.666234  0.613691  0.541635  0.902576   
2  0.879529  0.674238  1.000000  0.868409  0.922650  0.914616  0.653940   
3  0.747518  0.666234  0.868409  1.000000  0.823210  0.891044  0.718928   
4  0.770150  0.613691  0.922650  0.823210  1.000000  0.823958  0.604450   
5  0.862662  0.541635  0.914616  0.891044  0.823958  1.000000  0.616434   
6  0.446917  0.902576  0.653940  0.718928  0.604450  0.616434  1.000000   
7  0.742043  0.683105  0.840487  0.883104  0.867464  0.801593  0.663532   
8  0.757836  0.558006  0.790964  0.744359  0.799954  0.772431  0.545015   

          7         8  
0  0.742043  0.757836  
1  0.683105  0.558006  
2  0.840487  0.790964  
3  0.883104  0.744359  
4  0.867464  0.799954  
5  0.801593  0.772431  
6  0.663532  0.545015  
7  1.000000  0.775686  
8  0.775686  1.000000

Sample covariance matrix of weekly log returns:

$$
\begin{pmatrix}
0.002113 & 0.0009 & 0.001855 & 0.001289 & 0.000978 & 0.001765 & 0.000573 & 0.000779 & 0.000846 \\
0.0009 & 0.001742 & 0.001295 & 0.001044 & 0.000714 & 0.001025 & 0.001045 & 0.000656 & 0.00056 \\
0.001855 & 0.001295 & 0.002103 & 0.001494 & 0.001173 & 0.001879 & 0.000843 & 0.000887 & 0.000886 \\
0.001289 & 0.001044 & 0.001494 & 0.0014 & 0.000855 & 0.00149 & 0.000749 & 0.000761 & 0.00068 \\
0.000978 & 0.000714 & 0.001173 & 0.000855 & 0.000774 & 0.001021 & 0.00047 & 0.000557 & 0.000544 \\
0.001765 & 0.001025 & 0.001879 & 0.00149 & 0.001021 & 0.001995 & 0.000774 & 0.000822 & 0.000841 \\
0.000573 & 0.001045 & 0.000843 & 0.000749 & 0.00047 & 0.000774 & 0.000764 & 0.000426 & 0.000368 \\
0.000779 & 0.000656 & 0.000887 & 0.000761 & 0.000557 & 0.000822 & 0.000426 & 0.000532 & 0.000435 \\
0.000846 & 0.00056 & 0.000886 & 0.00068 & 0.000544 & 0.000841 & 0.000368 & 0.000435 & 0.000594
\end{pmatrix}
$$

In [39]:
print("Sample covariance matrix of weekly log returns:")
log_week_cov = pd.DataFrame(cov_log(weekly))
log_week_cov

Sample covariance matrix of weekly log returns:


0         1         2         3         4         5         6  \
0  0.002113  0.000900  0.001855  0.001289  0.000978  0.001765  0.000573   
1  0.000900  0.001742  0.001295  0.001044  0.000714  0.001025  0.001045   
2  0.001855  0.001295  0.002103  0.001494  0.001173  0.001879  0.000843   
3  0.001289  0.001044  0.001494  0.001400  0.000855  0.001490  0.000749   
4  0.000978  0.000714  0.001173  0.000855  0.000774  0.001021  0.000470   
5  0.001765  0.001025  0.001879  0.001490  0.001021  0.001995  0.000774   
6  0.000573  0.001045  0.000843  0.000749  0.000470  0.000774  0.000764   
7  0.000779  0.000656  0.000887  0.000761  0.000557  0.000822  0.000426   
8  0.000846  0.000560  0.000886  0.000680  0.000544  0.000841  0.000368   

          7         8  
0  0.000779  0.000846  
1  0.000656  0.000560  
2  0.000887  0.000886  
3  0.000761  0.000680  
4  0.000557  0.000544  
5  0.000822  0.000841  
6  0.000426  0.000368  
7  0.000532  0.000435  
8  0.000435  0.000594

Sample correlation matrix of weekly log returns:

$$
\begin{pmatrix}
1.0 & 0.469213 & 0.879804 & 0.749428 & 0.76497 & 0.859807 & 0.450752 & 0.734511 & 0.75526 \\
0.469213 & 1.0 & 0.676502 & 0.668763 & 0.615029 & 0.550044 & 0.905788 & 0.681816 & 0.550721 \\
0.879804 & 0.676502 & 1.0 & 0.870898 & 0.919974 & 0.917388 & 0.664763 & 0.838013 & 0.792214 \\
0.749428 & 0.668763 & 0.870898 & 1.0 & 0.821866 & 0.891789 & 0.724464 & 0.882196 & 0.745531 \\
0.76497 & 0.615029 & 0.919974 & 0.821866 & 1.0 & 0.822116 & 0.611227 & 0.868739 & 0.801646 \\
0.859807 & 0.550044 & 0.917388 & 0.891789 & 0.822116 & 1.0 & 0.62695 & 0.797574 & 0.771984 \\
0.450752 & 0.905788 & 0.664763 & 0.724464 & 0.611227 & 0.62695 & 1.0 & 0.667419 & 0.546301 \\
0.734511 & 0.681816 & 0.838013 & 0.882196 & 0.868739 & 0.797574 & 0.667419 & 1.0 & 0.773278 \\
0.75526 & 0.550721 & 0.792214 & 0.745531 & 0.801646 & 0.771984 & 0.546301 & 0.773278 & 1.0
\end{pmatrix}
$$

In [40]:
print("Sample correlation matrix of weekly log returns:")
log_week_corr = pd.DataFrame(corr_log(weekly))
log_week_corr

Sample correlation matrix of weekly log returns:


0         1         2         3         4         5         6  \
0  1.000000  0.469213  0.879804  0.749428  0.764970  0.859807  0.450752   
1  0.469213  1.000000  0.676502  0.668763  0.615029  0.550044  0.905788   
2  0.879804  0.676502  1.000000  0.870898  0.919974  0.917388  0.664763   
3  0.749428  0.668763  0.870898  1.000000  0.821866  0.891789  0.724464   
4  0.764970  0.615029  0.919974  0.821866  1.000000  0.822116  0.611227   
5  0.859807  0.550044  0.917388  0.891789  0.822116  1.000000  0.626950   
6  0.450752  0.905788  0.664763  0.724464  0.611227  0.626950  1.000000   
7  0.734511  0.681816  0.838013  0.882196  0.868739  0.797574  0.667419   
8  0.755260  0.550721  0.792214  0.745531  0.801646  0.771984  0.546301   

          7         8  
0  0.734511  0.755260  
1  0.681816  0.550721  
2  0.838013  0.792214  
3  0.882196  0.745531  
4  0.868739  0.801646  
5  0.797574  0.771984  
6  0.667419  0.546301  
7  1.000000  0.773278  
8  0.773278  1.000000

**Comparisons**

- The covariance matrix's values of weekly log returns are smaller than those of percentage returns, which means the variation of log returns is smaller than that of percentage returns
- The first statement does not hold for sample correlation matrices, since correlation is scaled using the standard deviation of each variable

(iii) Compute the sample covariance matrix of the monthly percentage returns of the
indeces, and the corresponding sample corelation matrix.

Compute the sample covariance and correlation matrices for monthly log returns,
and compare them with the corresponding matrices for monthly percentage returns.

Sample covariance matrix of monthly percentage returns:

$$
\begin{pmatrix}
0.00885 & 0.005215 & 0.007352 & 0.0057 & 0.004835 & 0.00551 & 0.002489 & 0.004447 & 0.004074 \\
0.005215 & 0.010158 & 0.005048 & 0.003909 & 0.002658 & 0.001627 & 0.004655 & 0.003242 & 0.002022 \\
0.007352 & 0.005048 & 0.006867 & 0.004682 & 0.004384 & 0.004468 & 0.002421 & 0.003867 & 0.003592 \\
0.0057 & 0.003909 & 0.004682 & 0.003901 & 0.002966 & 0.003762 & 0.001875 & 0.002934 & 0.002428 \\
0.004835 & 0.002658 & 0.004384 & 0.002966 & 0.002896 & 0.002943 & 0.001263 & 0.002452 & 0.002423 \\
0.00551 & 0.001627 & 0.004468 & 0.003762 & 0.002943 & 0.004555 & 0.000783 & 0.002678 & 0.002347 \\
0.002489 & 0.004655 & 0.002421 & 0.001875 & 0.001263 & 0.000783 & 0.002426 & 0.001644 & 0.001061 \\
0.004447 & 0.003242 & 0.003867 & 0.002934 & 0.002452 & 0.002678 & 0.001644 & 0.002352 & 0.002058 \\
0.004074 & 0.002022 & 0.003592 & 0.002428 & 0.002423 & 0.002347 & 0.001061 & 0.002058 & 0.002112
\end{pmatrix}
$$

In [41]:
monthly = value.resample("ME").last()
print("Sample covariance matrix of monthly percentage returns:")
pct_month_cov = pd.DataFrame(cov_pct(monthly))
pct_month_cov

Sample covariance matrix of monthly percentage returns:


0         1         2         3         4         5         6  \
0  0.008850  0.005215  0.007352  0.005700  0.004835  0.005510  0.002489   
1  0.005215  0.010158  0.005048  0.003909  0.002658  0.001627  0.004655   
2  0.007352  0.005048  0.006867  0.004682  0.004384  0.004468  0.002421   
3  0.005700  0.003909  0.004682  0.003901  0.002966  0.003762  0.001875   
4  0.004835  0.002658  0.004384  0.002966  0.002896  0.002943  0.001263   
5  0.005510  0.001627  0.004468  0.003762  0.002943  0.004555  0.000783   
6  0.002489  0.004655  0.002421  0.001875  0.001263  0.000783  0.002426   
7  0.004447  0.003242  0.003867  0.002934  0.002452  0.002678  0.001644   
8  0.004074  0.002022  0.003592  0.002428  0.002423  0.002347  0.001061   

          7         8  
0  0.004447  0.004074  
1  0.003242  0.002022  
2  0.003867  0.003592  
3  0.002934  0.002428  
4  0.002452  0.002423  
5  0.002678  0.002347  
6  0.001644  0.001061  
7  0.002352  0.002058  
8  0.002058  0.002112

Sample correlation matrix of monthly percentage returns:

$$
\begin{pmatrix}
1.0 & 0.549961 & 0.943076 & 0.970124 & 0.95498 & 0.867778 & 0.537142 & 0.974807 & 0.942299 \\
0.549961 & 1.0 & 0.604378 & 0.620937 & 0.490081 & 0.239122 & 0.937724 & 0.663404 & 0.436542 \\
0.943076 & 0.604378 & 1.0 & 0.904589 & 0.983171 & 0.798992 & 0.593265 & 0.962246 & 0.943275 \\
0.970124 & 0.620937 & 0.904589 & 1.0 & 0.882392 & 0.892545 & 0.609491 & 0.968568 & 0.845865 \\
0.95498 & 0.490081 & 0.983171 & 0.882392 & 1.0 & 0.810234 & 0.476495 & 0.93973 & 0.979739 \\
0.867778 & 0.239122 & 0.798992 & 0.892545 & 0.810234 & 1.0 & 0.235619 & 0.818322 & 0.756643 \\
0.537142 & 0.937724 & 0.593265 & 0.609491 & 0.476495 & 0.235619 & 1.0 & 0.688214 & 0.468561 \\
0.974807 & 0.663404 & 0.962246 & 0.968568 & 0.93973 & 0.818322 & 0.688214 & 1.0 & 0.92367 \\
0.942299 & 0.436542 & 0.943275 & 0.845865 & 0.979739 & 0.756643 & 0.468561 & 0.92367 & 1.0
\end{pmatrix}
$$

In [44]:
print("Sample correlation matrix of monthly percentage returns:")
pct_month_corr = pd.DataFrame(corr_pct(monthly))
pct_month_corr

Sample correlation matrix of monthly percentage returns:


0         1         2         3         4         5         6  \
0  1.000000  0.549961  0.943076  0.970124  0.954980  0.867778  0.537142   
1  0.549961  1.000000  0.604378  0.620937  0.490081  0.239122  0.937724   
2  0.943076  0.604378  1.000000  0.904589  0.983171  0.798992  0.593265   
3  0.970124  0.620937  0.904589  1.000000  0.882392  0.892545  0.609491   
4  0.954980  0.490081  0.983171  0.882392  1.000000  0.810234  0.476495   
5  0.867778  0.239122  0.798992  0.892545  0.810234  1.000000  0.235619   
6  0.537142  0.937724  0.593265  0.609491  0.476495  0.235619  1.000000   
7  0.974807  0.663404  0.962246  0.968568  0.939730  0.818322  0.688214   
8  0.942299  0.436542  0.943275  0.845865  0.979739  0.756643  0.468561   

          7         8  
0  0.974807  0.942299  
1  0.663404  0.436542  
2  0.962246  0.943275  
3  0.968568  0.845865  
4  0.939730  0.979739  
5  0.818322  0.756643  
6  0.688214  0.468561  
7  1.000000  0.923670  
8  0.923670  1.000000

Sample covariance matrix of monthly log returns:

$$
\begin{pmatrix}
0.009131 & 0.004831 & 0.007539 & 0.005811 & 0.004922 & 0.005484 & 0.002379 & 0.004494 & 0.004239 \\
0.004831 & 0.00877 & 0.004638 & 0.003679 & 0.002429 & 0.00161 & 0.004097 & 0.00296 & 0.001855 \\
0.007539 & 0.004638 & 0.006891 & 0.004761 & 0.004386 & 0.00444 & 0.002308 & 0.003874 & 0.003691 \\
0.005811 & 0.003679 & 0.004761 & 0.003941 & 0.002996 & 0.003699 & 0.001812 & 0.002943 & 0.00251 \\
0.004922 & 0.002429 & 0.004386 & 0.002996 & 0.002877 & 0.002898 & 0.001199 & 0.002441 & 0.002464 \\
0.005484 & 0.00161 & 0.00444 & 0.003699 & 0.002898 & 0.004311 & 0.000786 & 0.002633 & 0.002375 \\
0.002379 & 0.004097 & 0.002308 & 0.001812 & 0.001199 & 0.000786 & 0.002217 & 0.001559 & 0.001028 \\
0.004494 & 0.00296 & 0.003874 & 0.002943 & 0.002441 & 0.002633 & 0.001559 & 0.002334 & 0.002097 \\
0.004239 & 0.001855 & 0.003691 & 0.00251 & 0.002464 & 0.002375 & 0.001028 & 0.002097 & 0.002192
\end{pmatrix}
$$

In [47]:
print("Sample covariance matrix of monthly log returns:")
log_month_cov = pd.DataFrame(cov_log(monthly))
log_month_cov

Sample covariance matrix of monthly log returns:


0         1         2         3         4         5         6  \
0  0.009131  0.004831  0.007539  0.005811  0.004922  0.005484  0.002379   
1  0.004831  0.008770  0.004638  0.003679  0.002429  0.001610  0.004097   
2  0.007539  0.004638  0.006891  0.004761  0.004386  0.004440  0.002308   
3  0.005811  0.003679  0.004761  0.003941  0.002996  0.003699  0.001812   
4  0.004922  0.002429  0.004386  0.002996  0.002877  0.002898  0.001199   
5  0.005484  0.001610  0.004440  0.003699  0.002898  0.004311  0.000786   
6  0.002379  0.004097  0.002308  0.001812  0.001199  0.000786  0.002217   
7  0.004494  0.002960  0.003874  0.002943  0.002441  0.002633  0.001559   
8  0.004239  0.001855  0.003691  0.002510  0.002464  0.002375  0.001028   

          7         8  
0  0.004494  0.004239  
1  0.002960  0.001855  
2  0.003874  0.003691  
3  0.002943  0.002510  
4  0.002441  0.002464  
5  0.002633  0.002375  
6  0.001559  0.001028  
7  0.002334  0.002097  
8  0.002097  0.002192

Sample correlation matrix of monthly log returns:

$$
\begin{pmatrix}
1.0 & 0.539883 & 0.950377 & 0.968718 & 0.960392 & 0.874066 & 0.528889 & 0.973369 & 0.947534 \\
0.539883 & 1.0 & 0.596625 & 0.625677 & 0.483531 & 0.261835 & 0.929212 & 0.654228 & 0.423146 \\
0.950377 & 0.596625 & 1.0 & 0.913638 & 0.98497 & 0.814652 & 0.590508 & 0.966026 & 0.949549 \\
0.968718 & 0.625677 & 0.913638 & 1.0 & 0.889587 & 0.897343 & 0.612968 & 0.970428 & 0.853792 \\
0.960392 & 0.483531 & 0.98497 & 0.889587 & 1.0 & 0.822758 & 0.474618 & 0.941939 & 0.980916 \\
0.874066 & 0.261835 & 0.814652 & 0.897343 & 0.822758 & 1.0 & 0.254362 & 0.830165 & 0.772551 \\
0.528889 & 0.929212 & 0.590508 & 0.612968 & 0.474618 & 0.254362 & 1.0 & 0.685373 & 0.466092 \\
0.973369 & 0.654228 & 0.966026 & 0.970428 & 0.941939 & 0.830165 & 0.685373 & 1.0 & 0.927153 \\
0.947534 & 0.423146 & 0.949549 & 0.853792 & 0.980916 & 0.772551 & 0.466092 & 0.927153 & 1.0
\end{pmatrix}
$$

In [48]:
print("Sample correlation matrix of monthly log returns:")
log_month_corr = pd.DataFrame(corr_log(monthly))
log_month_corr

Sample correlation matrix of monthly log returns:


0         1         2         3         4         5         6  \
0  1.000000  0.539883  0.950377  0.968718  0.960392  0.874066  0.528889   
1  0.539883  1.000000  0.596625  0.625677  0.483531  0.261835  0.929212   
2  0.950377  0.596625  1.000000  0.913638  0.984970  0.814652  0.590508   
3  0.968718  0.625677  0.913638  1.000000  0.889587  0.897343  0.612968   
4  0.960392  0.483531  0.984970  0.889587  1.000000  0.822758  0.474618   
5  0.874066  0.261835  0.814652  0.897343  0.822758  1.000000  0.254362   
6  0.528889  0.929212  0.590508  0.612968  0.474618  0.254362  1.000000   
7  0.973369  0.654228  0.966026  0.970428  0.941939  0.830165  0.685373   
8  0.947534  0.423146  0.949549  0.853792  0.980916  0.772551  0.466092   

          7         8  
0  0.973369  0.947534  
1  0.654228  0.423146  
2  0.966026  0.949549  
3  0.970428  0.853792  
4  0.941939  0.980916  
5  0.830165  0.772551  
6  0.685373  0.466092  
7  1.000000  0.927153  
8  0.927153  1.000000

**Comparisons**

- The differences between covariance matrices and correlation matrices of monthly percentage returns and log returns are very close to each other

(iv) Comment on the differences between the sample covariance and correlation matrices for daily, weekly, and monthly returns.

- All entries of sample covariance matrix for monthly returns are larger than those for weekly returns


In [49]:
pct_month_cov - pct_week_cov

0         1         2         3         4         5         6  \
0  0.006736  0.004299  0.005483  0.004411  0.003840  0.003739  0.001917   
1  0.004299  0.008383  0.003735  0.002856  0.001932  0.000608  0.003597   
2  0.005483  0.003735  0.004733  0.003177  0.003187  0.002583  0.001581   
3  0.004411  0.002856  0.003177  0.002495  0.002099  0.002271  0.001125   
4  0.003840  0.001932  0.003187  0.002099  0.002107  0.001910  0.000791   
5  0.003739  0.000608  0.002583  0.002271  0.001910  0.002563  0.000018   
6  0.001917  0.003597  0.001581  0.001125  0.000791  0.000018  0.001652   
7  0.003652  0.002572  0.002963  0.002162  0.001885  0.001845  0.001214   
8  0.003222  0.001447  0.002699  0.001745  0.001874  0.001504  0.000690   

          7         8  
0  0.003652  0.003222  
1  0.002572  0.001447  
2  0.002963  0.002699  
3  0.002162  0.001745  
4  0.001885  0.001874  
5  0.001845  0.001504  
6  0.001214  0.000690  
7  0.001809  0.001617  
8  0.001617  0.001514

- All entries of sample covariance matrix for weekly returns are larger than those for daily returns

In [50]:
pct_week_cov - pct_day_cov

0         1         2         3         4         5         6  \
0  0.001584  0.000670  0.001348  0.000915  0.000618  0.001274  0.000376   
1  0.000670  0.001472  0.001038  0.000841  0.000516  0.000753  0.000841   
2  0.001348  0.001038  0.001560  0.001105  0.000796  0.001353  0.000617   
3  0.000915  0.000841  0.001105  0.001057  0.000574  0.001080  0.000576   
4  0.000618  0.000516  0.000796  0.000574  0.000463  0.000646  0.000306   
5  0.001274  0.000753  0.001353  0.001080  0.000646  0.001410  0.000543   
6  0.000376  0.000841  0.000617  0.000576  0.000306  0.000543  0.000581   
7  0.000501  0.000501  0.000595  0.000533  0.000323  0.000524  0.000291   
8  0.000557  0.000411  0.000581  0.000451  0.000314  0.000537  0.000235   

          7         8  
0  0.000501  0.000557  
1  0.000501  0.000411  
2  0.000595  0.000581  
3  0.000533  0.000451  
4  0.000323  0.000314  
5  0.000524  0.000537  
6  0.000291  0.000235  
7  0.000320  0.000245  
8  0.000245  0.000372

- The differences of correlation matrices for monthly and weekly are trivial, only with exceptions on the second (BNS) and fourth (GS) securities.

In [51]:
pct_month_corr - pct_week_corr

0         1             2         3             4             5  \
0  1.110223e-16  0.077269  6.354686e-02  0.222606  1.848301e-01  5.115514e-03   
1  7.726875e-02  0.000000 -6.985945e-02 -0.045297 -1.236102e-01 -3.025133e-01   
2  6.354686e-02 -0.069859 -1.110223e-16  0.036180  6.052164e-02 -1.156246e-01   
3  2.226060e-01 -0.045297  3.618017e-02  0.000000  5.918259e-02  1.500974e-03   
4  1.848301e-01 -0.123610  6.052164e-02  0.059183 -2.220446e-16 -1.372472e-02   
5  5.115514e-03 -0.302513 -1.156246e-01  0.001501 -1.372472e-02  2.220446e-16   
6  9.022504e-02  0.035149 -6.067534e-02 -0.109437 -1.279555e-01 -3.808148e-01   
7  2.327642e-01 -0.019702  1.217600e-01  0.085465  7.226588e-02  1.672854e-02   
8  1.844625e-01 -0.121464  1.523106e-01  0.101506  1.797854e-01 -1.578727e-02   

              6             7             8  
0  9.022504e-02  2.327642e-01  1.844625e-01  
1  3.514857e-02 -1.970166e-02 -1.214642e-01  
2 -6.067534e-02  1.217600e-01  1.523106e-01  
3 -1.094366e-01  8.546452e-02  1.015060e-01  
4 -1.279555e-01  7.226588e-02  1.797854e-01  
5 -3.808148e-01  1.672854e-02 -1.578727e-02  
6  2.220446e-16  2.468153e-02 -7.645425e-02  
7  2.468153e-02 -3.330669e-16  1.479838e-01  
8 -7.645425e-02  1.479838e-01 -2.220446e-16

- The differences of covariance matrices for weekly and daily are trivial, only with exceptions on the first (BAC), fourth (GS), and ninth (WFC) securities.

In [52]:
pct_week_corr - pct_day_corr

0             1             2         3             4             5  \
0  0.000000 -1.413912e-01 -6.405303e-02 -0.122325 -1.349186e-01 -3.062699e-02   
1 -0.141391  1.110223e-16  1.740070e-02  0.016241 -5.532774e-02 -8.978910e-02   
2 -0.064053  1.740070e-02  1.110223e-16 -0.023197 -5.557575e-03 -6.883288e-03   
3 -0.122325  1.624066e-02 -2.319679e-02  0.000000 -4.610599e-02 -2.092739e-02   
4 -0.134919 -5.532774e-02 -5.557575e-03 -0.046106  2.220446e-16 -6.459165e-02   
5 -0.030627 -8.978910e-02 -6.883288e-03 -0.020927 -6.459165e-02  1.110223e-16   
6 -0.165396  6.975755e-03 -1.840802e-02  0.049166 -5.797006e-02 -4.658560e-02   
7 -0.111178  3.162475e-02 -2.499738e-02  0.029233 -4.170616e-02 -5.912007e-02   
8 -0.094171 -6.712806e-02 -7.736786e-02 -0.079767 -6.576591e-02 -7.006800e-02   

              6             7         8  
0 -1.653963e-01 -1.111785e-01 -0.094171  
1  6.975755e-03  3.162475e-02 -0.067128  
2 -1.840802e-02 -2.499738e-02 -0.077368  
3  4.916646e-02  2.923325e-02 -0.079767  
4 -5.797006e-02 -4.170616e-02 -0.065766  
5 -4.658560e-02 -5.912007e-02 -0.070068  
6  2.220446e-16 -6.631285e-03 -0.102516  
7 -6.631285e-03  3.330669e-16 -0.103528  
8 -1.025162e-01 -1.035280e-01  0.000000

- The above results hold for both percentage returns and log returns.

## Problem 12

Consider a market made of cash and an asset with spot price $\$ 40$. The continuously compounded risk-free rate is $2 \%$. In four months, the price of the stock will be either $\$ 45$ or $\$ 35$.

(i) What is the payoff matrix $M$ in four months?

(ii) Is the market complete, i.e., is the matrix $M$ nonsingular?

(iii) How do you replicate a four months at-the-money call option on this asset, using the cash and the underlying asset?

*Solutions:*

(i) The value of $1 in 4 months is $1 \cdot e^{rt} = e^{0.02 \times \frac{4}{12}} =e^{0.02/3}$

Payoff matrix

$$M = \begin{pmatrix} e^{0.02/3} & e^{0.02/3} \\ 45 & 35 \end{pmatrix}$$

(ii) $\det(M) = e^{0.02/3}\times (35 - 45) = -10.066889 \neq 0$

$\therefore M$ is nonsingular.

(iii) The value of call option

$$C(T) = \max(S(T) - K, 0) = \begin{cases} S(T) - K & \text{if } S(T) > K \\ 0 & \text{if } S(T) \leq K \end{cases}$$

$$C(1/3) = \max(S(1/3) - K, 0)$$

state $w^1$: $S(1/3) = 45$, $C(1/3) = 45 - 40 = 5$

state $w^2$: $S(1/3) = 35$, $C(1/3) = 0$

∴ $C_{1/3} = \begin{pmatrix} 5 & 0 \end{pmatrix}$

Denote the positions of cash and asset of the replicated portfolio is $\Theta = (\theta_1, \theta_2)$

$$(\theta_1, \theta_2) \begin{pmatrix} e^{0.02/3} & e^{0.02/3} \\ 45 & 35 \end{pmatrix} = (5, 0)$$

∴ $\Theta M = C_{1/3} \rightarrow M^t \Theta^t = C_{1/3}^t$

$\begin{pmatrix} e^{0.02/3} & 45\\e^{0.02/3} & 35 \end{pmatrix} \begin{pmatrix} \theta_1 \\ \theta_2 \end{pmatrix} = \begin{pmatrix} 5 \\ 0 \end{pmatrix}$

$$
\begin{split}
\begin{pmatrix} \theta_1 \\ \theta_2 \end{pmatrix} &= \begin{pmatrix} e^{0.02/3} & 45\\e^{0.02/3} & 35 \end{pmatrix}^{-1} \begin{pmatrix} 5 \\ 0 \end{pmatrix}\\
&= \frac{1}{e^{0.02/3} \times (35 - 45)} \begin{pmatrix} 35 & -45\\ -e^{0.02/3}& e^{0.02/3} \end{pmatrix} \begin{pmatrix} 5 \\ 0 \end{pmatrix}\\
&= \begin{pmatrix} -17.383721 \\ 0.5 \end{pmatrix}
\end{split}
$$

The portfolio replicating the four months ATM call option on the asset is made of a long cash position of $17.3835 and a long 0.5 position in the asset.


In [ ]:
# Calculate for the last equation
x=1/(np.exp(0.02/3)*(-10)) * np.array([[35,-45],[-np.exp(0.02/3), np.exp(0.02/3)]]) @ np.array([[5], [0]])
print(x)

[[-17.383721]
 [  0.5     ]]


## Problem 13

In six months, the price of an asset with spot price $\$ 42$ will be either $\$ 35, \$ 38, \$ 40$, $\$ 43, \$ 45$, or $\$ 50$. Consider a market made of cash, the stock, and a six months at-themoney call and a six months at-the-money put option on one share of the stock. The risk-free interest rate is constant equal to $2 \%$.

(i) What is the payoff matrix in six months?

(ii) Is this market complete?

(iii) Are all the securities in this market non-redundant?

*Solutions:*

(i) 
$$
C(T) = \max(S(T) - K, 0) \\
P(T) = \max(K - S(T), 0)
$$

State $w^1$: $S_T = 35$ $C_T = 0$ $P_T = 7$

State $w^2$: $S_T = 38$ $C_T = 0$ $P_T = 4$ 

State $w^3$: $S_T = 40$ $C_T = 0$ $P_T = 2$ 

State $w^4$: $S_T = 43$ $C_T = 1$ $P_T = 0$ 

State $w^5$: $S_T = 45$ $C_T = 3$ $P_T = 0$ 

State $w^6$: $S_T = 50$ $C_T = 8$ $P_T = 0$

The value of \\$1 in six months would be $e^{0.02 \times \frac{1}{2}} = e^{0.01}$

Payoff matrix

$$
M = \begin{pmatrix}
e^{0.01} & e^{0.01} & e^{0.01} & e^{0.01} & e^{0.01} & e^{0.01} \\
35 & 38 & 40 & 43 & 45 & 50 \\
0 & 0 & 0 & 1 & 3 & 8 \\
7 & 4 & 2 & 0 & 0 & 0
\end{pmatrix}
$$

(ii) No, because there're fewer securities than market states, so the payoff matrix is singular. The market is incomplete.

(iii)

$$
\begin{pmatrix}
r_1 \\
r_2 + r_4 - r_3 \\
r_3 \\
r_4
\end{pmatrix}
=
\begin{pmatrix}
e^{0.01} & e^{0.01} & e^{0.01} & e^{0.01} & e^{0.01} & e^{0.01} \\
42 & 42 & 42 & 42 & 42 & 42 \\
0 & 0 & 0 & 1 & 3 & 8 \\
7 & 4 & 2 & 0 & 0 & 0
\end{pmatrix}
$$

$r_1$ is dependent on $r_2 + r_4 - r_3$. Each row represent an asset in six months $S_{i, 1/2}, i=1:4$

$$S_{2,1/2} + S_{4,1/2} - S_{3,1/2} = \frac{42}{e^{0.01}} S_{1,1/2}$$

i.e. 

$$
\begin{aligned}
S_{1,1/2} &= \frac{e^{0.01}}{42} (S_{2,1/2} + S_{4,1/2} - S_{3,1/2})\\
&=0.024049\cdot (S_{2,1/2} + S_{4,1/2} - S_{3,1/2})
\end{aligned}
$$

Therefore cash is a redundant security.

**Reference**:

- A Linear Algebra Primer for Financial Engineering
- [Backward error in GE; pivoting; iterative refinement](https://www.cs.cornell.edu/~bindel/class/cs6210-f09/lec12.pdf)